man/bnetd.1
.\"
.\" Copyright (C) 1998  Mark Baysinger (mbaysing@ucsd.edu)
.\" Copyright (C) 1998,1999  Ross Combs (ross@bnetd.org)
.\" Copyright (C) 1999  Rob Crittenden (rcrit@greyoak.com)
.\" 
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111,
.\" USA.
.\"
.TH BNETD 1 "12 June, 2000" "BNETD" "The bnetd Project"
.SH NAME
bnetd \- Unix Battle.net daemon
.SH SYNOPSIS
.B bnetd
[\fB-f\fP]
[\fB-c\fP \fIconfigfile\fP]
[\fB-d\fP \fIdumpfile\fP]
.SH DESCRIPTION
.B bnetd
is a program that will eventually completely emulate a Blizzard Battle.net
server.  Currently Starcraft, Brood War, Diablo, and Warcraft II BNE are
supported as clients.
.LP
It is known to work under most versions of Unix and can also be used under
Windows as a win32 console application.
.SH OPTIONS
.TP
.B \-c --config
Use this configuration file instead of the compiled-in default location.
.TP
.B \-d --hexdump
Put a hex dump of packets sent and received into the specified file.
.TP
.B \-f --foreground
The server normally acts as a daemon and goes into the background.  This forces
the server to run in the foreground.
.TP
.B \-h --help --usage
Print server usage information and exit.
.TP
.B \-v --version
Print its version number and exit.
.SH FILES
.TP
.I bnetd.conf
The main configuration file.  Among other things, it specifies the locations for
the other files used by the daemon.
.TP
.I bnmotd.txt
The "message of the day" file.  This is displayed when the user enters the
first chat channel.
.TP
.I bnetd_default_user
This is the default account file.  The contains all default account attributes
which are used when an attribute is not set on an actual account.

.I channel.list
This is the permanent channel list.  Short names, target clients, and other
attributes may be specified here.
.TP
.I news.txt
The contensts of this file are printed when the /news command is used.
.TP
.I ad.list
This is the ad banner list which determines which banners to display at which
times.  The order of display for the banners may be specified as well as the
URL which they should follow when clicked.
.TP
.I bnban
The IP ban file can be used to block connections from certain addresses.
.TP
.I bnhelp
This file contains the the chat command help.
.TP
.I gametrans
The game address translation file can be used to allow players from behind a
NAT firewall to play with those outside of the firewall.
.TP
.I autoupdate
This is the autoupdate MPQ archive list.  It chooses which versions of which
clients to update.
.TP
.I tos.txt
This is the Terms Of Service file that is displayed before a user can create
a new account.
.TP
.I icons.bni
The icons which the client displays for players in the chat rooms.
.TP
.I ad000001.pcx
This is the ad Banner. Currently, it says "Welcome to... BNETD".
.SH NOTES
The current version is available at:
.LP
.RS
.I http://www.bnetd.org/
.RE
.SH "SEE ALSO"
.BR bntrackd (1),
.BR bnproxy (1),
.BR bnchat (1),
.BR bnbot (1),
.BR bnpass (1),
.BR bnftp (1),
.BR bnetd.conf (5),
.BR bnmotd.txt (5)
.SH AUTHORS
Ross Combs (\fIross@bnetd.org\fP),
Rob Crittenden (\fIrob@bnetd.org\fP), and
Mark Baysinger (\fImark@bnetd.org\fP)


NEWS

		    PvPGN 1.7.x Major Changes And New Features
		  ==============================================

 o Westwood Online (WOL) support

    This is a historic moment in the life of the PvPGN project. For the first 
    time a different gaming network than Battle.net is supported with PvPGN. 
    Current supports includes almost complete support for Tiberian Sun (TS) 
    and basic support for Red Alert 2 (RA2) and Yuri's Revenge (YR).

 o WIN32 GUI for d2cs and d2dbs

    d2cs and d2bs now feature a new slick GUI if build under WIN32.

 o Instant startup (partial)

    PvPGN now can startup in a lot lesser time than before, this is due to 
    code changes which make it not load all accounts at startup. This 
    improvement should be mostly noticed on very high number of accounts 
    servers. Note that currently this happens only to SQL storage types 
    (the file based storage types still load all accounts at startup).

 o Optimized account loading/saving codes

    Account loading/saving codes have been rewritten allowing PvPGN to scale 
    a lot better with a high number of online users (ie lower CPU usage).

 o Better game result calculation

    Game result decision has been completly changed to use a democratic 
    ellection scheme (ie it bases the decision on the oppinions of the most 
    results). This should solve problems with game results beeing wrong under 
    some situations (like team games).

 o New WIN32 compiler/enviroment support

    Now you can compile PvPGN under WIN32 using complete GPL/free build tools 
    by using Dev-C++ project and additional tools (mingw, etc).

 o Optimized handling of D2 clients

    Replaced a O(n) search algorithm through the client connection list used 
    mostly by D2 handling functions with a O(1) algorithm.

 o Optimized timers codes

    Timers codes have been rewritten so the most used functions complete a 
    lot faster than before. This improves PvPGN CPU usage a lot because some 
    timers functions are used very often inside PvPGN.

 o Memory management changes and optimizations

    There has been added some memory management wrappers functions which 
    never fail and thus the calling codes were simplified a lot by removing 
    checks to memory allocation failures. If memory allocation fails PvPGN 
    tries to recover by programming an immediate shutdown and cleaning up 
    some safe memory.

 o Changed codes to use logged in username instead of account username

    This change makes PvPGN use the username introduced at login (with the 
    exact case as in login) instead of the account name. This solves some 
    bugs (like could not edit profile in D2 if logged in with different case) 
    and also gets us more close to Battle.Net.

 o Switched from string based tags to integer based tags

    Codes have been dramatically changed to use integers for storing of tags 
    (client tags, architecture tags, etc) and thus when comparing tags we use 
    plain integer comparation instead of string comparation. This should 
    reduce CPU usage a lot.

 o Optimized list handling codes

    The list API has been changed and the codes have been optimized so it 
    will remove elements from the list upon request and not later, thus 
    eliminating a O(n) search and thus making PvPGN a lot faster.

 o Unified and extended address translation system

    We have unified all address translation files and codes to use a single 
    address_translation.conf file for this job. Also the file syntax has 
    been completly changed to be more flexible and allow configuration of 
    almost any thinkable NAT setup.

 o IRC support reworked

    IRC client support has been bugfixed, optimized and made to work with a 
    lot more IRC clients.

 o Unified file based storage

    File based storage (plain and cdb) have been unified into a single storage 
    backend (called "file") and the mode (plain/cdb) is just a parameter.

 o Rewritten low level socket multiplexing codes (fdwatch2)

    The codes which deal with how PvPGN knows which sockets are available for 
    reading/writing have been rewritten with a new interface which does not 
    presume (like the old codes) that the underlying OS allocates socket 
    values "close" (ex. WIN32 systems do not while most Unix based systems do 
    so). The new codes eliminates this presumtion and thus allows WIN32 hosts 
    to support a large number of connections as PvPGN supported Unix based 
    platforms.

 o Added more storage types: SQLite3 and ODBC

    With 1.7.x PvPGN now supports besides the MySQL/PostgreSQL SQL backends 
    also SQLite3 (www.sqlite.org) and ... ODBC SQL engine !! This means that 
    theoretically now you should be able to use PvPGN with almost any SQL 
    engine as long as an ODBC driver exists for it.
 

src/d2cs/bnetd.cpp
/*
 * Copyright (C) 2000,2001	Onlyer	(onlyer@263.net)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
#include "common/setup_before.h"
#include "setup.h"
#include "bnetd.h"

#include <ctime>

#include "common/eventlog.h"
#include "prefs.h"
#include "s2s.h"
#include "handle_bnetd.h"
#include "common/setup_after.h"

namespace pvpgn
{

	namespace d2cs
	{

		static t_connection * bnetd_connection = NULL;

		extern int bnetd_init(void)
		{
			return bnetd_check();
		}

		extern int bnetd_check(void)
		{
			static unsigned int	prev_connecting_checktime = 0;

			if (bnetd_connection) {
				if (d2cs_conn_get_state(bnetd_connection) == conn_state_connecting) {
					if (std::time(NULL) - prev_connecting_checktime > prefs_get_s2s_timeout()) {
						eventlog(eventlog_level_warn, __FUNCTION__, "connection to bnetd s2s timeout");
						d2cs_conn_set_state(bnetd_connection, conn_state_destroy);
						return -1;
					}
				}
				return 0;
			}
			if (!(bnetd_connection = s2s_create(prefs_get_bnetdaddr(), BNETD_SERV_PORT, conn_class_bnetd))) {
				return -1;
			}
			if (d2cs_conn_get_state(bnetd_connection) == conn_state_init) {
				handle_bnetd_init(bnetd_connection);
			}
			else {
				prev_connecting_checktime = std::time(NULL);
			}
			return 0;
		}


		extern t_connection * bnetd_conn(void)
		{
			return bnetd_connection;
		}

		extern int bnetd_set_connection(t_connection * c)
		{
			bnetd_connection = c;
			return 0;
		}

		extern int bnetd_destroy(t_connection * c)
		{
			if (bnetd_connection != c) {
				eventlog(eventlog_level_error, __FUNCTION__, "bnetd connection do not match");
				return -1;
			}
			bnetd_connection = NULL;
			return 0;
		}

	}

}
